core/src/cubos/core/data/ser/json.cpp

```diff
@@ -0,0 +1,128 @@
+#include <cubos/core/log.hpp>
+
+/// [JSON serializer]
+#include <cubos/core/data/ser/json.hpp>
+
+using cubos::core::data::JSONSerializer;
+using cubos::core::reflection::Type;
+
+/// [MACRO to add hooks for primitive types]
+#define AUTO_HOOK(type, ...)                                                                                           \
+    this->hook<type>([this](const type& value) {                                                                       \
+        mJSON = value;                                                                                                 \
+        return true;                                                                                                   \
+    })
+
+/// [Setting up hooks]
+#include <cubos/core/reflection/external/primitives.hpp>
+#include <cubos/core/reflection/external/string.hpp>
+
+using cubos::core::reflection::reflect;
+
+JSONSerializer::JSONSerializer()
+{
+    AUTO_HOOK(bool);
+    AUTO_HOOK(char);
+    AUTO_HOOK(int8_t);
+    AUTO_HOOK(int16_t);
+    AUTO_HOOK(int32_t);
+    AUTO_HOOK(int64_t);
+    AUTO_HOOK(uint8_t);
+    AUTO_HOOK(uint16_t);
+    AUTO_HOOK(uint32_t);
+    AUTO_HOOK(uint64_t);
+    AUTO_HOOK(float);
+    AUTO_HOOK(double);
+    AUTO_HOOK(std::string);
+}
+/// [Setting up hooks]
+
+nlohmann::json JSONSerializer::output()
+{
+    return mJSON;
+}
+
+/// [Decomposing types]
+#include <cubos/core/reflection/traits/array.hpp>
+#include <cubos/core/reflection/traits/dictionary.hpp>
+#include <cubos/core/reflection/traits/fields.hpp>
+#include <cubos/core/reflection/traits/string_conversion.hpp>
+#include <cubos/core/reflection/type.hpp>
+
+using cubos::core::reflection::ArrayTrait;
+using cubos::core::reflection::DictionaryTrait;
+using cubos::core::reflection::FieldsTrait;
+using cubos::core::reflection::StringConversionTrait;
+
+bool JSONSerializer::decompose(const Type& type, const void* value)
+{
+    if (type.has<ArrayTrait>())
+    {
+        auto jsonArr = nlohmann::json::array();
+
+        const auto& arrayTrait = type.get<ArrayTrait>();
+
+        for (const auto* element : arrayTrait.view(value))
+        {
+            if (!this->write(arrayTrait.elementType(), element))
+            {
+                CUBOS_WARN("Could not deserialize array element");
+                return false;
+            }
+            jsonArr.push_back(mJSON);
+        }
+
+        mJSON = jsonArr;
+        return true;
+    }
+
+    if (type.has<StringConversionTrait>())
+    {
+        const auto& trait = type.get<StringConversionTrait>();
+        mJSON = trait.into(value);
+    }
```

Comment:
```suggestion
        return true;
    }
```

Reply:
I don't get this change. It also fails the uuid test:
`CHECK( (json.dump()) == ("\"f7063cd4-de44-47b5-b0a9-f0e7a558c9e5\"") ) is NOT correct!
  values: CHECK( "Hello, world!" == "f7063cd4-de44-47b5-b0a9-f0e7a558c9e5" )`

Reply:
nvm got it